plugins/Utils/windowstatestorage.cpp

```diff
@@ -16,167 +17,329 @@
 
 #include "windowstatestorage.h"
 
-#include <QtConcurrent>
 #include <QDebug>
+#include <QDir>
+#include <QMetaObject>
+#include <QObject>
 #include <QSqlQuery>
 #include <QSqlError>
 #include <QSqlResult>
+#include <QStandardPaths>
 #include <QRect>
 #include <unity/shell/application/ApplicationInfoInterface.h>
 
-QMutex WindowStateStorage::s_mutex;
 
-inline QString sanitiseString(QString string) {
-    return string.remove(QLatin1Char('\"'))
-                 .remove(QLatin1Char('\''))
-                 .remove(QLatin1Char('\\'));
-}
-
-WindowStateStorage::WindowStateStorage(const QString& dbName, QObject *parent):
-    QObject(parent)
+class AsyncQuery: public QObject
 {
-    const QString dbPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + QStringLiteral("/unity8/");
-    m_db = QSqlDatabase::addDatabase(QStringLiteral("QSQLITE"));
-    QDir dir;
-    dir.mkpath(dbPath);
-    if (dbName != nullptr) {
-        m_db.setDatabaseName(dbName);
-    } else {
-        m_db.setDatabaseName(dbPath + "windowstatestorage.sqlite");
+    Q_OBJECT
+
+public:
+    AsyncQuery(const QString& dbName):
+        m_dbName(dbName)
+    {
     }
-    initdb();
-}
 
-WindowStateStorage::~WindowStateStorage()
-{
-    m_db.close();
-}
+    ~AsyncQuery()
+    {
+        QSqlDatabase::removeDatabase(m_connectionName);
+    }
 
-void WindowStateStorage::saveState(const QString &windowId, WindowStateStorage::WindowState state)
-{
-    const QString queryString = QStringLiteral("INSERT OR REPLACE INTO state (windowId, state) values ('%1', '%2');")
-            .arg(sanitiseString(windowId))
-            .arg((int)state);
+    Q_PROPERTY (const QString dbName READ getDbName)
```

Comment:
minor: typically properties are declared right below the `Q_OBJECT` macro (whether it's private or public, it does not matter)